avocado/utils/asset.py

```diff
@@ -71,21 +71,178 @@ def __init__(self, name, asset_hash, algorithm, locations, cache_dirs,
-        else:
+
+        if algorithm:
             self.algorithm = algorithm
+        else:
+            self.algorithm = DEFAULT_HASH_ALGORITHM
 
+        self.locations = []
+        # If `name` is an url, move it to `locations`
+        if parsed_url.scheme:
+            self.locations.append(os.path.dirname(parsed_url.geturl()))
+        # Let's be conservative and check `locations` type
         if isinstance(locations, str):
-            self.locations = [locations]
+            # remove asset name from location
+            if self.basename in locations:
+                parsed_locations = urlparse.urlparse(locations)
+                self.locations.append(
+                    os.path.dirname(parsed_locations.geturl()))
+            else:
+                self.locations.append(locations)
+        elif isinstance(locations, list):
+            # remove asset name from location
+            for location in locations:
+                if self.basename in location:
+                    parsed_location = urlparse.urlparse(location)
+                    self.locations.append(
+                        os.path.dirname(parsed_location.geturl()))
+                else:
+                    self.locations.append(location)
+
+        # Let's make the user's life easy and handle the cache
+        if cache_dirs:
+            self.cache_dirs = cache_dirs
         else:
-            self.locations = locations
-        self.cache_dirs = cache_dirs
+            # Break early if we don't have the cache_dirs. This is necessary
+            # so we don't change the Asset signature.
+            raise EnvironmentError("At least one cache directory is necessary"
+                                   " to download the asset.")
+
         self.expire = expire
         self.metadata = metadata
 
+        # Break early. We need to know the source location of the asset.
+        if not self.locations:
+            raise EnvironmentError("An asset should have at least one source"
+                                   " location. Double check your asset name"
+                                   " and location and make sure to use a"
+                                   " supported protocol.")
+
+    def _create_hash_file(self, asset_path):
+        """
+        Compute the hash of the asset file and add it to the CHECKSUM
+        file.
+
+        :param asset_file: full path of the asset file.
+        """
+        result = crypto.hash_file(asset_path, algorithm=self.algorithm)
+        with open(self._get_hash_file(asset_path), 'w') as hash_file:
+            hash_file.write('%s %s\n' % (self.algorithm, result))
+
+    def _create_metadata_file(self, asset_dir):
+        """
+        Creates JSON file with metadata.
+        The file will be saved as `asset_file`_metadata.json
+
+        :param asset_dir: full path where the asset file is located.
+        """
+        metadata_file_name = "%s_metadata.json" % self.basename
```

Comment:
Would make sense to define an `_get_metadata_file` just like `_get_hash_file`?

Reply:
There is a `get_metadata` on line 387. I left it public for future extension.

Reply:
Ah, I see what you mean. Note taken!